### PR TITLE
Batchnorm.py - Type Annotations Support

### PR DESCRIPTION
**Pull request by technicalgrp89**
tuna/miopen/driver/batchnorm.py
-- type hints
-- mypy compatibility
-- utility.groovy to execute the mypy cmd

### CODE REVIEW

File: tuna/miopen/driver/batchnorm.py [hunk 1]
```diff
@@ -125,27 +130,27 @@ def get_bn_dict(self):
...
   @staticmethod
-  def get_check_valid(tok1, tok2):
+  def get_check_valid(tok1: str, tok2: str) -> str:
     """Check if valid BN arg"""
     return arg_valid(tok1, tok2)
 
-  def get_db_obj(self, keep_id=False):
+  def get_db_obj(self, keep_id: bool = False) -> dict:
     """Return the DB representation of this object"""
     return BNConfig(**self.compose_tensors(keep_id))
```
Comment (cderb) [inline tuna/miopen/driver/batchnorm.py]: BNConfig is not a dict

File: tuna/miopen/driver/batchnorm.py [hunk 2]
```diff
@@ -125,27 +130,27 @@ def get_bn_dict(self):
...
-  def get_params(tok1):
+  def get_params(tok1: str) -> str:
     """Get full arg name"""
     return get_fd_name(tok1, TABLE_COLS_BN_MAP)
 
   @staticmethod
-  def get_check_valid(tok1, tok2):
+  def get_check_valid(tok1: str, tok2: str) -> str:
     """Check if valid BN arg"""
     return arg_valid(tok1, tok2)
```
Comment (cderb) [inline tuna/miopen/driver/batchnorm.py]: arg_valid returns a bool

File: tuna/miopen/driver/batchnorm.py [hunk 3]
```diff
@@ -94,18 +99,18 @@ def compute_direction(self):
       raise ValueError("Can't import driver commmand line, \
           one and only one of forw or back must be set")
 
-  def parse_row(self, db_obj):
+  def parse_row(self, db_obj: BNConfig) -> None:
     """Overwritting base class function for batch_norm"""
     return self.parse_bn_row(db_obj)
```
Comment (cderb) [inline tuna/miopen/driver/batchnorm.py]: Remove return.

**Review by cderb — CHANGES_REQUESTED** [review]

File: tuna/miopen/driver/batchnorm.py [hunk 4]
```diff
@@ -42,49 +44,52 @@
...
+    self.back: int = 0
+    self.mode: int = 0
+    self.run: int = 0
+    self.in_d: int = 1
+    self.in_h: int = 1
+    self.in_w: int = 1
+    self.in_channels: int = 1
+    self.in_layout: str = 'NCHW'
+    self.num_dims: Optional[str] = None
+    self.direction: Union[str, int, None] = None
```
Comment (cderb) [inline tuna/miopen/driver/batchnorm.py]: This won't be optional, they will also be a singular type. Consult the defaults.

File: tuna/miopen/driver/batchnorm.py [hunk 5]
```diff
@@ -125,27 +130,27 @@ def get_bn_dict(self):
...
 
   @staticmethod
-  def get_params(tok1):
+  def get_params(tok1: str) -> str:
     """Get full arg name"""
     return get_fd_name(tok1, TABLE_COLS_BN_MAP)
 
   @staticmethod
-  def get_check_valid(tok1, tok2):
+  def get_check_valid(tok1: str, tok2: str) -> Callable[[str, str], bool]:
```
Comment (cderb) [inline tuna/miopen/driver/batchnorm.py]: This does not return a callable.

**Review by cderb — CHANGES_REQUESTED** [review]

File: tuna/miopen/driver/batchnorm.py [hunk 6]
```diff
@@ -42,49 +44,52 @@
...
+    self.back: int = 0
+    self.mode: int = 0
+    self.run: int = 0
+    self.in_d: int = 1
+    self.in_h: int = 1
+    self.in_w: int = 1
+    self.in_channels: int = 1
+    self.in_layout: str = 'NCHW'
+    self.num_dims: int = 2
+    self.direction: Union[str, int] = 'F'
```
Comment (cderb) [inline tuna/miopen/driver/batchnorm.py]: only string

File: tuna/miopen/driver/batchnorm.py [hunk 7]
```diff
@@ -42,49 +44,52 @@
...
-  def parse_driver_line(self, line):
+  def parse_driver_line(self, line: str) -> None:
     super().parse_driver_line(line)
 
     self.compute_direction()
 
-  def compute_direction(self):
+  def compute_direction(self) -> None:
     """Setting BN direction based on forw and back"""
     self.direction = int(self.forw) + 4 * int(self.back)
```
Comment (cderb) [inline tuna/miopen/driver/batchnorm.py]: Change this self.direction (int) to a function local variable to avoid affecting the typing of self.direction.

**Review by cderb — CHANGES_REQUESTED** [review]

**Review by cderb — APPROVED** [review]

**Review by alexandraBara — APPROVED** [review]

**Pull request merged**